Security

Sadymix opened the pull request.

Commit:
added enum UserPermission with 3 fields and internal variable that has getter

Commit:
added dependencies for spring security and google guava

Commit:
added User class, which overrides the methods from UserDetails

Commit:
added selectUserByUsername method

Commit:
added UserDao field
added constructor
added loadUserByUserName, which returns the user or throw an exception

Commit:
added:
PasswordEncoder field
constructor
selectUserByUsername method, which returns first user found by username
getUsers method which returns List with two users

Commit:
added bean passwordEncoder
returns BCryptPasswordEncoder - 10 strength of encryption

Commit:
Enum with 2 fields
added constructor, getter
added getGrantedAuthorities, which returns permissions

Commit:
moved User to models package
removed unused methods (isAccountNonExpired(), isAccountNonLocked(), isCredentialsNonExpired(), isEnabled() - to @Data annotation)
changed getAuthorities to returning empty List

Commit:
deleted auth packaged with UserDao.java and UserDaoService.java

Commit:
remove enum UserRole, added UserRole entity class

Commit:
removed security package with PasswordConfig.java and UserPermission.java

Commit:
moved UserService.java to services package
removed @SupressWarnings
changed UserDao to UserRepo
modified loadUserByUserName method

Commit:
added WebSecurityConfig
added @Autowire DataSourceField
added overrided configure method for http configuration
added configureGlobal method for auth configuration

Commit:
added UserRepo that extends CrudRepository
added findByUsername method

Commit:
added UserServiceTest unit test
added static final User field (builder)
mocked user
added setup method - stubbed userRepo
added testLoadUserByUsername method, which verifies number of use findByUsername method and checks that username is the same after loadUserByUsername method

Commit:
added setUp method for MockMvc
modified testGetProductList method (basic auth), added @WithMockUser annotation
added @WithMockUser to testGetCartProductList

Commit:
removed unnecessary properties

Commit:
added application.yml to the tests with h2 db

Commit:
added dependency for the h2 db

Commit:
removed @WebMvcTest
added @RunWith and @SpringBootTest
changed @WithMockitUser to default values for both methods
added accept(APPLICATION_JSON_VALUE) to mockMVC

Commit:
removed @WebMvcTest
added @RunWith and @SpringBootTest
changed @WithMockitUser to default values for both methods
added accept(APPLICATION_JSON_VALUE) to mockMVC

pszerszen commented on src/main/java/com/example/shopapp/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,57 @@
+public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Autowired
+    private DataSource dataSource;
```
Make the field final and create a constructor using Lombok

pszerszen commented on pom.xml:
```diff
@@ -1,90 +1,108 @@
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>31.0.1-jre</version>
+        </dependency>
```
I think this dependency is redundant, isn't it?

pszerszen commented on src/main/java/com/example/shopapp/services/UserService.java:
```diff
@@ -0,0 +1,24 @@
+    @Autowired
+    public UserService(UserRepo userRepo) {
+        this.userRepo = userRepo;
+    }
```
make the field final and create constructor using Lombok

pszerszen commented on src/main/java/com/example/shopapp/services/UserService.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        return userRepo.findByUsername(username);
```
Add some logic here to handle missing user case. If repository doesn't find a user with the username, throw `UsernameNotFoundException`. Of course add a proper unit test case.

pszerszen reviewed (changes requested):
Aside from the rest of the comments add integration test cases with unauthenticated requests.

Commit:
removed unused guava dependency

Commit:
changed to @Required args contructor instead of field dependency -> constructor dependency

Commit:
changed UserRepo from field dependency to constructor dependency using lombok @RequiredArgsConstructor
modified loadUserByUsername method to handle the user if its null and throws UserNotFoundException with message

Commit:
changed testLoadUserByUsername method to check if exception occurs for the message of UserNameNotFoundException

pszerszen commented on src/main/java/com/example/shopapp/services/UserService.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        var user = userRepo.findByUsername(username);
+        if (user == null) {
+            throw new UsernameNotFoundException(username);
+        } else {
+            return user;
+        }
```
```suggestion
        return Optional.ofNullable(userRepo.findByUsername(username))
                .orElseThrow(() -> new UsernameNotFoundException(username))
```

pszerszen commented on src/test/java/com/example/shopapp/services/UserServiceTest.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Test
+    void testLoadUserByUsername() {
+        var user = userService.loadUserByUsername("user");
+        Exception exception = assertThrows(UsernameNotFoundException.class, () -> {
+            userService.loadUserByUsername("admin");
+        });
+        verify(userRepo).findByUsername("user");
+        assertEquals("user", user.getUsername());
+        assertTrue(exception.getMessage().contains("admin"));
+    }
```
split it to 2 test methods

pszerszen reviewed (changes requested):
Integration test

Commit:
modified the loadUserByUsername body for better readability using Optional class

Commit:
split testLoadUserName to testLoadByUsernameExceptionOccurrence

Commit:
added new test case for checking the unauthorized request
